warehouse/schema/schema.go

```diff
@@ -67,44 +66,48 @@ func NewSchema(
 		warehouse:                        warehouse,
 		schemaRepo:                       repo.NewWHSchemas(db),
 		stagingFileRepo:                  repo.NewStagingFiles(db),
-		log:                              logger,
+		log:                              logger.Child("schema"),
 		stagingFilesSchemaPaginationSize: conf.GetInt("Warehouse.stagingFilesSchemaPaginationSize", 100),
 		skipDeepEqualSchemas:             conf.GetBool("Warehouse.skipDeepEqualSchemas", false),
 		enableIDResolution:               conf.GetBool("Warehouse.enableIDResolution", false),
 	}
 }
 
-func (sh *Schema) updateLocalSchema(ctx context.Context, uploadId int64, updatedSchema model.Schema) error {
-	_, err := sh.schemaRepo.Insert(ctx, &model.WHSchema{
-		UploadID:        uploadId,
-		SourceID:        sh.warehouse.Source.ID,
-		Namespace:       sh.warehouse.Namespace,
-		DestinationID:   sh.warehouse.Destination.ID,
-		DestinationType: sh.warehouse.Type,
-		Schema:          updatedSchema,
-	})
-	if err != nil {
-		return fmt.Errorf("updating local schema: %w", err)
+// ConsolidateStagingFilesUsingLocalSchema consolidates staging files schema with warehouse schema
+func (sh *Schema) ConsolidateStagingFilesUsingLocalSchema(ctx context.Context, stagingFiles []*model.StagingFile) (model.Schema, error) {
+	consolidatedSchema := model.Schema{}
+	batches := lo.Chunk(stagingFiles, sh.stagingFilesSchemaPaginationSize)
+	for _, batch := range batches {
+		schemas, err := sh.stagingFileRepo.GetSchemasByIDs(ctx, repo.StagingFileIDs(batch))
+		if err != nil {
+			return nil, fmt.Errorf("getting staging files schema: %v", err)
+		}
+
+		consolidatedSchema = consolidateStagingSchemas(consolidatedSchema, schemas)
 	}
 
-	sh.localSchema = updatedSchema
+	sh.localSchemaMu.RLock()
+	consolidatedSchema = consolidateWarehouseSchema(consolidatedSchema, sh.localSchema)
+	consolidatedSchema = overrideUsersWithIdentifiesSchema(consolidatedSchema, sh.warehouse.Type, sh.localSchema)
+	sh.localSchemaMu.RUnlock()
 
-	return nil
-}
+	consolidatedSchema = enhanceDiscardsSchema(consolidatedSchema, sh.warehouse.Type)
+	consolidatedSchema = enhanceSchemaWithIDResolution(consolidatedSchema, sh.isIDResolutionEnabled(), sh.warehouse.Type)
```

Comment:
Don't we need write lock here?

Reply:
The lock is to protect `localSchema` only here. `consolidatedSchema` is a brand new map and we're not copying any reference in it that might be coming from `localSchema` so we should be cool. Thoughts?

Reply:
Sorry, I missed this part. `consolidatedSchema` creates a brand new map.